test_runner/src/main/kotlin/ftl/filter/TestFilters.kt

```diff
@@ -0,0 +1,181 @@
+            notIgnored()
+        } else {
+            val parsedFilters =
+                targets
+                    .asSequence()
+                    .map(String::trim)
+                    .map(TestFilters::parseSingleFilter)
+                    .toList()
+
+            // select test method name filters and short circuit if they match ex: class a.b#c
+            val testMethod = parsedFilters.filter { it.isTestMethod() }.toTypedArray()
+            val otherFilters = parsedFilters.filterNot { it.isTestMethod() }.toTypedArray()
+
+            allOf(notIgnored(), anyOf(*testMethod, allOf(*otherFilters)))
+        }
+    }
+
+    private fun ITestFilter.isTestMethod(): Boolean {
+        return name.startsWith("withClassName") && name.contains("#")
+    }
+
+    private fun parseSingleFilter(target: String): ITestFilter {
+        val matcher = FILTER_ARGUMENT.matcher(target)
+        require(matcher.matches()) { "Invalid argument: $target" }
+        val args = matcher.group(2)
+            .split(",")
+            .map(String::trim)
+        val command = matcher.group(1)
+        return when (command) {
+            ARGUMENT_TEST_CLASS -> withClassName(args)
+            ARGUMENT_NOT_TEST_CLASS -> not(withClassName(args))
+            ARGUMENT_TEST_PACKAGE -> withPackageName(args)
+            ARGUMENT_NOT_TEST_PACKAGE -> not(withPackageName(args))
+            ARGUMENT_ANNOTATION -> withAnnotation(args)
+            ARGUMENT_NOT_ANNOTATION -> not(withAnnotation(args))
+            ARGUMENT_TEST_FILE -> fromTestFile(args)
+            ARGUMENT_NOT_TEST_FILE -> not(fromTestFile(args))
+            ARGUMENT_TEST_SIZE -> withSize(args)
+            else -> throw IllegalArgumentException("Filtering option $command not supported")
+        }
+    }
+
+    private fun withSize(args: List<String>): ITestFilter =
+        withAnnotation(args.map { Size.from(it).annotation })
+
+    private fun fromTestFile(args: List<String>): ITestFilter {
+        require(args.size == 1) { "Invalid file path" }
+        val path = Paths.get(args[0])
+        try {
+            val lines = Files.readAllLines(path)
+            // this is really an implementation detail:
+            // being the package name most generic one, it is able to filter properly if you pass the package name,
+            // the fully qualified class name or the fully qualified method name.
+            return withPackageName(lines)
+        } catch (e: IOException) {
+            throw RuntimeException("Unable to read testFile", e)
+        }
+    }
+
+    private fun withPackageName(packageNames: List<String>): ITestFilter = object : ITestFilter {
+        override val name = "withPackageName ${packageNames.joinToString(", ")}"
+        override fun run(testMethod: TestMethod): Boolean {
+            require(packageNames.isNotEmpty()) { MISSING_ARGUMENTS_MSG }
+            return packageNames.any { packageName -> testMethod.testName.startsWith(packageName) }
```

Comment:
Either all errors should be extracted as a constant or none. Do you actually think the error message here makes still sense?

Reply:
I inlined the message and added a few more error handling statements.

I think it makes sense to have sanity checks in the filters, that helps reduce the risk of an invalid test run.